dask_expr/shuffle.py

```diff
@@ -462,6 +461,55 @@ def _layer(self):
         return toolz.merge(dsk1, dsk2, dsk3, dsk4)
 
 
+class P2PShuffle(SimpleShuffle):
+    """P2P worker-based shuffle implementation"""
+
+    lazy_hash_support = False
+
+    def _layer(self):
+        from distributed.shuffle._shuffle import (
+            ShuffleId,
+            barrier_key,
+            shuffle_barrier,
+            shuffle_transfer,
+            shuffle_unpack,
+        )
+
+        dsk = {}
+        token = self._name.split("-")[-1]
+        _barrier_key = barrier_key(ShuffleId(token))
+        name = "shuffle-transfer-" + token
+        transfer_keys = list()
+        parts_out = (
+            self._partitions if self._filtered else list(range(self.npartitions_out))
+        )
+        for i in range(self.frame.npartitions):
+            transfer_keys.append((name, i))
+            dsk[(name, i)] = (
+                shuffle_transfer,
+                (self.frame._name, i),
+                token,
+                i,
+                self.npartitions_out,
+                self.partitioning_index,
+                parts_out,
+            )
+
+        dsk[_barrier_key] = (shuffle_barrier, token, transfer_keys)
+
+        # TODO: Decompose p2p Into transfer/barrier + unpack
+        name = self._name
+        for i, part_out in enumerate(parts_out):
```

Comment:
Is there a specific reason for not reusing the partition name/ID in the task name? I feel like this will add one additional indirection to keep in mind when debugging in the future.

Reply:
I'm not certain that I fully understand what you are saying here, but here is what I *think* you are looking for:

**Regarding the name**: We need the name of these final `shuffle_unpack` tasks to match the name of the expression (i.e. `self._name`).

**Regarding the index**: Since `P2PShuffle` is technically a `PartitionsFiltered` instance, we also need to make sure that the output indices are always a contiguous range starting with `0` (e.g. [0, 1, 2, 3]). That is, if `self._partitions` is `[1, 3, 4]`, it means we must remap what would **usually** be partitions `[1, 3, 4]` to have the indices `[0, 1, 2]`. This `PartitionsFiltered` mechanism is absolutely an unfortunate level of indirection. However, it essentially allows us to avoid low-level culling in dask-expr.

Reply:
Sorry, I referred to the use of the `i` as the index in the task key (not name). That explanation clears it up, I wasn't aware that a contiguous range is expected. Thanks!